ferry/transform/same_courses.py

```diff
@@ -2,21 +2,92 @@
 Find historical offerings of a course.
 """
 
-from typing import Any
+from typing import cast, Any
 
+import logging
 import itertools
 import edlib
-import networkx
+import networkx as nx
 import pandas as pd
 from tqdm import tqdm
-import logging
 
+# TODO: I don't quite like these hard-coded constants. Is there a better way
+# to measure similarity that's more context-aware?
 MIN_TITLE_MATCH_LEN = 8
 MIN_DESCRIPTION_MATCH_LEN = 32
 
 MAX_TITLE_DIST = 0.25
 MAX_DESCRIPTION_DIST = 0.25
 
+# These are courses that changed code; we only consider course pairs that have
+# overlapping codes because we can't afford a full pairwise text comparison.
+# These have to be added as special cases.
+# We do so by pretending that the specified course codes have an extra cross-
+# listing, so they can be grouped into the same shared-code component for further
+# title comparison.
+# TODO: is there a data structure that allows us to find similar texts quickly?
```

Comment:
Not sure if you mean speeding up the calculation of Levenshtein distance between 2 strings, but if so, this might be interesting to explore.

https://github.com/ashvardanian/StringZilla

Check out the Levenshtein distance performance comparison.

Reply:
I think we are not really stuck on a constant time, it's that we have to use some data structure so we don't do a quadratic pairwise text comparison.